journalpump/senders/base.py

```diff
@@ -192,29 +195,63 @@ def request_stop(self):
         self.running = False
 
     def send_messages(self, *, messages, cursor):
+        # This should be overridden in the classes that inherit this
         pass
 
     def maintenance_operations(self):
         # This can be overridden in the classes that inherit this
         pass
 
+    def send_heartbeat(self):
+        # This can be overridden in the classes that inherit this
+        pass
+
     def run(self):
+        self.log.info("Starting")
         while self.running:
-            try:
-                # Don't run maintenance operations again immediately if it just failed
-                if not self.last_maintenance_fail or time.monotonic() - self.last_maintenance_fail > 60:
-                    self.maintenance_operations()
-            except Exception as ex:  # pylint: disable=broad-except
-                self.log.error("Maintenance operation failed: %r", ex)
-                self.stats.unexpected_exception(ex=ex, where="maintenance_operation")
-                self.last_maintenance_fail = time.monotonic()
-            if len(self.msg_buffer) > 1000 or \
-               time.monotonic() - self.last_send_time > self.max_send_interval:
+            if self.should_perform_maintenance():
+                self.perform_maintenance()
+
+            if self.should_send_hearbeat():
+                self.do_send_heartbeat()
+
+            if self.should_send_messages():
                 self.get_and_send_messages()
-            else:
-                time.sleep(0.1)
+
+            time.sleep(0.1)
         self.log.info("Stopping")
 
+    def should_perform_maintenance(self):
+        # Don't run maintenance operations again immediately if it just failed
+        return not self.last_maintenance_fail or time.monotonic() - self.last_maintenance_fail > 60
+
+    def should_send_messages(self):
+        max_send_interval_exceeded = time.monotonic() - self.last_send_time > self.max_send_interval
+        buffer_size_over_send_threshold = len(self.msg_buffer) > 1000
+        return buffer_size_over_send_threshold or max_send_interval_exceeded
+
+    def should_send_hearbeat(self):
+        return self.max_heartbeat_interval and time.monotonic() - self.last_heartbeat_time > self.max_heartbeat_interval
+
+    def perform_maintenance(self):
+        self.log.info("Performing maintenance")
+        try:
+            self.maintenance_operations()
+        except Exception as ex:  # pylint: disable=broad-except
+            self.log.error("Maintenance operation failed: %r", ex)
+            self.stats.unexpected_exception(ex=ex, where="maintenance_operation")
+            self.last_maintenance_fail = time.monotonic()
+
+    def do_send_heartbeat(self):
+        self.log.info("Sending heartbeat message")
+        try:
+            self.send_heartbeat()
+        except Exception:  # pylint: disable=broad-except
+            self.log.exception("Problem sending heartbeat message")
+            self._backoff()
```

Comment:
backoff doesn't seem sensible for heartbeat, it's scheduled operation anyway.

Reply:
:heavy_check_mark: